### PR TITLE
Close pending channel when unexpected error is returned from PublishTransaction

### LINKED ISSUE #4760: wallet: respect policy-level transaction size limits when crafting regular and funding transactions
Tracking this here instead of the `btcwallet` issue for more visibility. As is, we don't always ensure that a transaction either we construct manually (for funding) or within `btcwallet` itself is below the current widely used policy that constraints transaction sizes. Thankfully, we can simply start using the existing policy level function within `btcd` itself to enforce these limits: https://github.com/btcsuite/btcd/blob/master/mempool/policy.go#L271 

When a user encounters this issue, they'll see this in the logs: 
```
: unmatched backend error: -26: 64: tx-size
```

And then end up with a dangling transaction that'll keep being rebroadcast, when instead we should remove it from the tx store.

### PR DESCRIPTION
**Pull request by yyforyongyu**
Fix #4760 

We will abort the funding flow and close the pending channel if the returned error from `PublishTransaction` is unexpected.

### CODE REVIEW

File: funding/manager.go [hunk 1]
```diff
@@ -1956,6 +1958,90 @@ func (f *Manager) handleFundingCreated(peer lnpeer.Peer,
...
+	// Return early if there's no error.
+	if err == nil {
+		return false
+	}
+
+	log.Errorf("Unable to broadcast funding tx %x for "+
+		"ChannelPoint(%v): %v", fundingTxBuf.Bytes(),
+		ch.FundingOutpoint, err)
+
+	// Note that the ErrDoubleSpend actually wraps two erros returned from
```
Comment (alexbosworth) [inline funding/manager.go:2019]:
```suggestion
	// Note that the ErrDoubleSpend actually wraps two errors returned from
```

**Comment by alexbosworth** [comment]
With external funding I've had a case where the publish had an error but then the channel did confirm anyway and everything worked out. This was due to a funding tx that was valid but had an unconfirmed parent that was pushed out of the local mempool due to fee conditions, making the child invalid

**Comment by yyforyongyu** [comment]
> With external funding I've had a case where the publish had an error but then the channel did confirm anyway and everything worked out. This was due to a funding tx that was valid but had an unconfirmed parent that was pushed out of the local mempool due to fee conditions, making the child invalid

Sounds like it's due to the [check here](https://github.com/btcsuite/btcd/blob/130ea5bddde33df32b06a1cdb42a6316eb73cff5/mempool/mempool.go#L1409) that causes the btcwallet to return an [`ErrDoubleSpend` here](https://github.com/btcsuite/btcwallet/blob/master/wallet/wallet.go#L3568). Let me try to patch an itest to this PR to address this case.

**Comment by halseth** [comment]
It could be dangerous to assume a publication error means the channel will never confirm. 

An alternative solution that could be more "safe" is one I started way back: https://github.com/lightningnetwork/lnd/pull/1755

Essentially the idea here was to make it possible to "release" the inputs used to the funding tx, and then wait for them to be spent before abandoning the channel.

**Comment by Crypt-iQ** [comment]
> It could be dangerous to assume a publication error means the channel will never confirm.

Not disagreeing, but under what circumstances can this occur?

**Comment by yyforyongyu** [comment]
> It could be dangerous to assume a publication error means the channel will never confirm.

Yeah agreed. I think one direction is to look deeper into what is causing the publication error, probably [here in btcwallet](https://github.com/btcsuite/btcwallet/blob/master/wallet/wallet.go#L3501) and [here in `bitcoind`](https://github.com/bitcoin/bitcoin/blob/4a540683ec40393d6369da1a9e02e45614db936d/src/policy/policy.h#L19). This means we need to decide explicitly to abort the funding flow or not based on each of the error.

The other solution you mentioned, if I got it right, which is to double spend the input then fails the channel. This is nice as we don't need to examine each of the errors. However what if the input was never spent again and the pending channel would stuck there forever?

**Comment by halseth** [comment]
> > It could be dangerous to assume a publication error means the channel will never confirm.
> 
> Not disagreeing, but under what circumstances can this occur?

Alex mentioned one above. Also in case of Neutrino we currently just look at any errors the peers sends us (tbh we should probably change this) to determine if publication succeeded. This means that the peer could easily trick us by sending a fake reject.

> However what if the input was never spent again and the pending channel would stuck there forever?

I think this is just a cosmetic edge case. If the input is never spent again then the use most likely isn't actively using the wallet.

**Comment by Crypt-iQ** [comment]
> 
> 
> With external funding I've had a case where the publish had an error but then the channel did confirm anyway and everything worked out. This was due to a funding tx that was valid but had an unconfirmed parent that was pushed out of the local mempool due to fee conditions, making the child invalid

how did the channel eventually confirm?

**Comment by alexbosworth** [comment]
> > With external funding I've had a case where the publish had an error but then the channel did confirm anyway and everything worked out. This was due to a funding tx that was valid but had an unconfirmed parent that was pushed out of the local mempool due to fee conditions, making the child invalid
> 
> how did the channel eventually confirm?

I'd imagine a case where you have an unconfirrmed parent, that isn't present on the local machine but then it does confirm, you restart lnd, this rebroadcasts, now your channel that failed broadcast locally initially is confirmed

**Comment by Crypt-iQ** [comment]
In Alex's scenario with this patch, the funding tx would not be broadcast and then the pending channel is failed. Which would be fine as rebroadcast would not be attempted on startup (and thus won't ever get into the mempool).

For neutrino, as Johan said, the peer could lie in `MsgReject` and relay the funding tx, where we would fail the pending channel completely.

For the PR that Johan linked, it would forget the channel if the funding tx's inputs are used.  But, if lnd has locked these inputs, it may not be possible to spend these inputs and forget the channel.

The linked issue is mostly concerned with policy-level checks.  I think we could do some sanity checks *before* the transaction is ever broadcast in `handleFundingSigned`.

**Comment by yyforyongyu** [comment]
!lightninglabs-deploy mute 2022-Feb-21

**Comment by yyforyongyu** [comment]
Replaced by #6400.

**Comment by saubyk** [comment]
Since this PR is replaced by https://github.com/lightningnetwork/lnd/pull/6400, removed the release milestone

**Pull request closed without merging**